OpenRA.Game/Support/ConditionExpression.cs

```diff
@@ -0,0 +1,901 @@
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							ast.Push(Expressions.Expression.LessThanOrEqual(x, y));
+							continue;
+						}
+
+						case TokenType.GreaterThan:
+						{
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							ast.Push(Expressions.Expression.GreaterThan(x, y));
+							continue;
+						}
+
+						case TokenType.GreaterThanOrEqual:
+						{
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							ast.Push(Expressions.Expression.GreaterThanOrEqual(x, y));
+							continue;
+						}
+
+						case TokenType.Add:
+						{
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							ast.Push(Expressions.Expression.Add(x, y));
+							continue;
+						}
+
+						case TokenType.Subtract:
+						{
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							ast.Push(Expressions.Expression.Subtract(x, y));
+							continue;
+						}
+
+						case TokenType.Multiply:
+						{
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							ast.Push(Expressions.Expression.Multiply(x, y));
+							continue;
+						}
+
+						case TokenType.Divide:
+						{
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							var isNotZero = Expressions.Expression.NotEqual(y, Zero);
+							var divide = Expressions.Expression.Divide(x, y);
+							ast.Push(IfThenElse(isNotZero, divide, MaxInt));
+							continue;
+						}
+
+						case TokenType.Modulo:
+						{
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							var isNotZero = Expressions.Expression.NotEqual(y, Zero);
+							var modulo = Expressions.Expression.Modulo(x, y);
+							ast.Push(IfThenElse(isNotZero, modulo, Zero));
+							continue;
+						}
+
+						case TokenType.False:
+						{
+							ast.Push(False);
+							continue;
+						}
+
+						case TokenType.True:
+						{
+							ast.Push(True);
+							continue;
+						}
+
+						case TokenType.Number:
+						{
+							ast.Push(Expressions.Expression.Constant(((NumberToken)t).Value));
+							continue;
+						}
+
+						case TokenType.Variable:
+						{
+							var symbol = Expressions.Expression.Constant(((VariableToken)t).Symbol);
+							Func<string, IReadOnlyDictionary<string, int>, int> parseSymbol = ParseSymbol;
+							ast.Push(Expressions.Expression.Call(parseSymbol.Method, symbol, SymbolsParam));
+							continue;
+						}
+
+						default:
+							throw new InvalidProgramException(
+								"ConditionExpression.Compiler.Compile() is missing an expression builder for TokenType.{0}".F(
+									Enum<TokenType>.GetValues()[(int)t.Type]));
+					}
+				}
+
+				return Expressions.Expression.Lambda<Func<IReadOnlyDictionary<string, int>, int>>(
+					ast.Pop(ExpressionType.Int), SymbolsParam).Compile();
```

Comment:
Providing a name for the compiled function would be handy for debugging purposes. A fixed string like "ConditionExpression" for a bare minimum of context would be nice. If you wanted to be fancy, maybe try passing the trait name (or some other such contextual strings) down to here so you can give them more detailed names.